File: python-implementation/huffman_tree.py
from struct import unpack

class HuffmanTreeNode:
    # Initializing the leaf node
    def __init__(self, length=None, val=None, internal=False):
        self.left = None
        self.right = None
        self.parent = None
        self.val = val
        self.internal = internal
        self.length = length
    
    def __str__(self):
        return f'The value is {self.val} and the length is {self.length}'

class HuffmanTree:
    # Data is the huffman table chunk 
    # showing the character lengths
    # Example: 0 1 5 1 1 1 1 1 1 0 0 0 0 0 0 a b c 
    # The passed data is the after 4 bytes of the section, behind the markers
    def __init__(self, data, root):
        self.data = data
        self.elements = self.create_elements(data)
        self.root = root
        self.codes = {}

    def create_elements(self, data):
        offset = 0
        header, = unpack("B",data[offset:offset+1])
        offset += 1

        # Extract the 16 bytes containing length data
        lengths = unpack("BBBBBBBBBBBBBBBB", data[offset:offset+16]) 
        offset = offset + 16

        elements = []
        for index, val in enumerate(lengths):
            for j in range(val):
                current_val = unpack("B", data[offset:offset+1]) # getting the character
                offset += 1 
                elements.append(HuffmanTreeNode(index + 1, current_val)) # we are using lengths and all leaf nodes have the characters

        # Sorting the elements in descending order in length --> ascending order in frequency
        return sorted(elements, key=lambda item: item.length)
    

    def add_to_tree(self, root, element, pos):
        if root.internal == False:
            return False
        # If position is one, we found the place to insert
        if pos == 1:
            if root.left == None:
                root.left = element
            elif root.right == None:
                root.right = element
            else:
                return False
            return True
        for i in [0,1]:
            internal_node = HuffmanTreeNode(val=i, internal=True)
            if i == 0:
                if root.left == None:
                    root.left = internal_node
                if self.add_to_tree(root.left, element, pos-1) == True:
                    return True
            else:
                if root.right == None:
                    root.right = internal_node
                if self.add_to_tree(root.right, element, pos-1) == True:
                    return True
        return False

    # Returns the codes for characters depending on the huffman tree.
    def decode_tree(self, cur_node, current_bit_string):
        # Base case reached to a leaf node
        if cur_node.val != None:
            self.codes[cur_node.val] = current_bit_string
        # Appending 0 when going left
        if cur_node.left != None:
            self.decode_tree(cur_node.left, current_bit_string + '0')
        # Appending 1 when going right
        if cur_node.right != None:
            self.decode_tree(cur_node.right, current_bit_string + '1')
        return

File: cuda2_1-implementation/src/huffmanTree.cpp
#include "huffmanTree.h"

// Constructor for HuffmanTree
HuffmanTree::HuffmanTree(const std::vector<uint8_t>& bytes) {
    this->bytes = bytes;
    this->createNodes();
    this->root = new HuffmanTreeNode(0,0,false);

    for(const auto& n: this->nodes) {
        this->addToTree(this->root, n, n->length);
    }
    this->decodeTree(this->root, "");
}

// Creates the nodes for the characters.
void HuffmanTree::createNodes() {
    // Extracting the length information.
    std::vector<char> lengths(this->bytes.begin(), this->bytes.begin() + 16);
    int offset = 16;

    for (int i = 0; i < 16; i++) {
        char curLength = lengths[i];
        for (int j = 0; j < curLength; j++) {
            char curVal = this->bytes[offset++];
            this->nodes.push_back(new HuffmanTreeNode(curVal, i+1, true));
        }
    }
}

bool HuffmanTree::addToTree(HuffmanTreeNode* root, HuffmanTreeNode* node, int position) {
    if (root->isLeaf){
        return false;
    }
    if (position == 1) {
        if (root->left == NULL) {
            root->left = node;
        } else if (root->right == NULL) {
            root->right = node;
        } else {
            return false;
        }
        return true;
    }
    std::vector<int> indeces = {0,1};
    for (const int& i: indeces) {
        HuffmanTreeNode* internalNode = new HuffmanTreeNode(i, 0, false);
        if (i == 0) {
            if (root->left == NULL) {
                root->left = internalNode;
            }
            if (this->addToTree(root->left, node, position-1)) {
                return true;
            }
        } else {
            if (root->right == NULL) {
                root->right = internalNode;
            }
            if (this->addToTree(root->right, node, position - 1)) {
                return true;
            }
        }
    }
    return false;
}

void HuffmanTree::decodeTree(HuffmanTreeNode* node, std::string currentString) {
    // If we reach a leaf node we return the value.
    if (node->isLeaf) {
        this->codes[node->val] = currentString;
        return;
    }
    if (node->left != NULL) {
        this->decodeTree(node->left, currentString + "0");
    }
    if (node->right != NULL) {
        this->decodeTree(node->right, currentString + "1");
    }
    return;
}

uint8_t HuffmanTree::traverseTree(HuffmanTreeNode* cur, Stream* st) {
    if (cur->isLeaf) {
        return cur->val;
    }
    int val = (int)(st->getBit());

    if ((cur->right != NULL) && val == 1) {
        return this->traverseTree(cur->right, st);
    } else if ((cur->left != NULL) && val == 0) {
        return this->traverseTree(cur->left, st);
    }

    return cur->val;
}

uint8_t HuffmanTree::getCode(Stream* st) {
    return this->traverseTree(this->root, st);
}
